### PR TITLE
[NEW] Leave Teams

### PR DESCRIPTION
**Pull request by gerzonc**
## Proposed changes

With this PR we add the functionality for leaving teams which includes the possibility of leaving teams channels' in the process.

Depends on #3090 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves a current function)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/views/SelectListView.js [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+		);
+	}
+
+	renderChannel = ({
+		onPress, testID, title, icon, checked, alert
+	}) => {
+		const { theme } = this.props;
+
+		return (
+			<Touch
```
Comment (diegolmello) [inline app/views/SelectListView.js:181]: List.Item

**Review by diegolmello — CHANGES_REQUESTED** [review]

File: app/views/RoomActionsView/index.js [hunk 2]
```diff
@@ -412,6 +414,79 @@ class RoomActionsView extends React.Component {
...
 
+	handleLeaveTeam = async(selected) => {
+		const { room } = this.state;
+		try {
+			const { navigation, isMasterDetail } = this.props;
+			const result = await RocketChat.leaveTeam({ teamName: room.name });
+
+			if (selected) {
+				try {
+					selected.map(item => RocketChat.leaveRoom(item.rid, item.t));
```
Comment (diegolmello) [inline app/views/RoomActionsView/index.js:425]: `teams.leave` accepts a `rooms` option that does it already.
https://github.com/RocketChat/Rocket.Chat/blob/develop/app/api/server/v1/teams.ts#L316

File: app/views/RoomActionsView/index.js [hunk 3]
```diff
@@ -412,6 +414,79 @@ class RoomActionsView extends React.Component {
...
+			const db = database.active;
+			const subCollection = db.get('subscriptions');
+			const teamChannels = await subCollection.query(
+				Q.where('team_id', room.teamId),
+				Q.where('team_main', null)
+			);
+
+			if (teamChannels.length) {
+				navigation.navigate('SelectListView', {
+					title: 'Leave_Team', room, data: teamChannels, subtitle: 'Select_Teams', nextAction: data => this.handleLeaveTeam(data)
```
Comment (diegolmello) [inline app/views/RoomActionsView/index.js:466]: `handleLeaveTeam` param is too vague.
Here it receives `data`, which seems to be an Array.
Below in the code, it receives `room.name`, which is a String.

File: app/views/SelectListView.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
+/* eslint-disable no-mixed-spaces-and-tabs */
```
Comment (diegolmello) [inline app/views/SelectListView.js:1]: Why?

File: app/views/SelectListView.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+	renderList = () => {
+		const { data } = this.state;
+		const { theme } = this.props;
+
+		return (
+			<FlatList
+				data={data}
+				extraData={this.state}
+				keyExtractor={item => item._id}
```
Comment (diegolmello) [inline app/views/SelectListView.js:154]: Why `_id`?

File: app/views/SelectListView.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+		);
+	}
+
+	render() {
+		const { loading } = this.state;
+
+		return (
+			<SafeAreaView testID='new-message-view'>
+				<StatusBar />
+				{this.renderList()}
```
Comment (diegolmello) [inline app/views/SelectListView.js:170]: Can you move the `FlatList` here instead of calling this function?

File: app/views/SelectListView.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+				keyboardShouldPersistTaps='always'
+			/>
+		);
+	}
+
+	render() {
+		const { loading } = this.state;
+
+		return (
+			<SafeAreaView testID='new-message-view'>
```
Comment (diegolmello) [inline app/views/SelectListView.js:168]: Copy/pasted testid

File: app/views/SelectListView.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+		}),
+		theme: PropTypes.string,
+		isMasterDetail: PropTypes.bool
+	};
+
+	constructor(props) {
+		super(props);
+		const data = props.route?.params?.data;
+		this.title = props.route?.params?.title;
+		this.teamName = props.route?.params?.teamName;
```
Comment (diegolmello) [inline app/views/SelectListView.js:46]: What does this do?

File: app/views/SelectListView.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+		const options = {
+			headerShown: true,
+			headerTitleAlign: 'center',
+			headerTitle: I18n.t(this.title)
+		};
+
+		if (isMasterDetail) {
+			options.headerLeft = () => <HeaderButton.CloseModal navigation={navigation} />;
+		} else {
+			options.headerLeft = () => <HeaderBackButton labelVisible={false} onPress={() => navigation.pop()} tintColor={themes[theme].headerTintColor} />;
```
Comment (diegolmello) [inline app/views/SelectListView.js:69]: You did the same thing on a previous PR.
This is a default behaviour.

File: app/views/SelectListView.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+		);
+
+		navigation.setOptions(options);
+	}
+
+	renderHeader = () => {
+		const { theme } = this.props;
+		return (
+			<View style={{ backgroundColor: themes[theme].backgroundColor }}>
+				<Text style={[styles.buttonText, { color: themes[theme].bodyText }]}>{I18n.t('Select_Teams')}</Text>
```
Comment (diegolmello) [inline app/views/SelectListView.js:85]: This is a reusable component. Why `select teams`?

File: app/views/SelectListView.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+			</View>
+		);
+	}
+
+	showAlert = () => {
+		Alert.alert(
+			I18n.t('Cannot_leave'),
+			I18n.t('Last_owner_team_room'),
```
Comment (diegolmello) [inline app/views/SelectListView.js:93]: Why?

File: app/views/SelectListView.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+			</>
+		);
+	}
+
+	renderList = () => {
+		const { data } = this.state;
+		const { theme } = this.props;
+
+		return (
+			<FlatList
```
Comment (diegolmello) [inline app/views/SelectListView.js:151]: Missing separator on first item

File: app/views/SelectListView.js [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+import StatusBar from '../containers/StatusBar';
+import { themes } from '../constants/colors';
+import { withTheme } from '../theme';
+import SafeAreaView from '../containers/SafeAreaView';
+import { animateNextTransition } from '../utils/layoutAnimation';
+import Loading from '../containers/Loading';
+
+const styles = StyleSheet.create({
+	buttonText: {
+		fontSize: 17,
```
Comment (diegolmello) [inline app/views/SelectListView.js:24]: It's 16 on Figma.

File: app/views/SelectListView.js [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+		margin: 16,
+		...sharedStyles.textRegular
+	}
+});
+
+class SelectListView extends React.Component {
+	static propTypes = {
+		navigation: PropTypes.object,
+		route: PropTypes.object,
+		user: PropTypes.shape({
```
Comment (diegolmello) [inline app/views/SelectListView.js:34]: Not used

**Review by diegolmello — CHANGES_REQUESTED** [review]

File: app/views/SelectListView.js (hunk 4 above)
Comment (gerzonc) [reply app/views/SelectListView.js:1]: Removed

File: app/views/SelectListView.js (hunk 5 above)
Comment (gerzonc) [reply app/views/SelectListView.js:154]: Updated to `item.rid`

File: app/views/SelectListView.js (hunk 8 above)
Comment (gerzonc) [reply app/views/SelectListView.js:46]: Previously used for the leaveTeam function, no needed here anymore

File: app/views/SelectListView.js (hunk 10 above)
Comment (gerzonc) [reply app/views/SelectListView.js:85]: Solved

File: app/views/SelectListView.js (hunk 11 above)
Comment (gerzonc) [reply app/views/SelectListView.js:93]: Moved to `RoomActionsView`

File: app/views/RoomActionsView/index.js (hunk 3 above)
Comment (gerzonc) [reply app/views/RoomActionsView/index.js:466]: Fixed

File: app/views/RoomActionsView/index.js [hunk 15]
```diff
@@ -412,6 +414,85 @@ class RoomActionsView extends React.Component {
...
 
+	handleLeaveTeam = async(selected) => {
+		try {
+			const { room } = this.state;
+			const { navigation, isMasterDetail } = this.props;
+			const result = await RocketChat.leaveTeam({ teamName: room.name, ...(selected && { rooms: selected }) });
+
+			if (result.success) {
+				if (isMasterDetail) {
+					navigation.popToTop();
```
Comment (diegolmello) [inline app/views/RoomActionsView/index.js:425]: Why not `Navigation.navigate('DrawerNavigator');`?

File: app/views/RoomActionsView/index.js [hunk 16]
```diff
@@ -412,6 +414,85 @@ class RoomActionsView extends React.Component {
...
+				} else {
+					navigation.navigate('RoomsListView');
+				}
+			}
+		} catch (e) {
+			log(e);
+			Alert.alert(
```
Comment (diegolmello) [inline app/views/RoomActionsView/index.js:432]: Can you use `showErrorAlert` instead?

File: app/views/RoomActionsView/index.js [hunk 17]
```diff
@@ -412,6 +414,85 @@ class RoomActionsView extends React.Component {
...
+						text: 'OK',
+						style: 'cancel'
+					}
+				]
+			);
+		}
+	}
+
+	showNoLeaveTeamAlert = () => {
+		Alert.alert(
```
Comment (diegolmello) [inline app/views/RoomActionsView/index.js:446]: `showErrorAlert`

File: app/views/RoomActionsView/index.js [hunk 18]
```diff
@@ -412,6 +414,85 @@ class RoomActionsView extends React.Component {
...
+				Q.where('team_id', room.teamId),
+				Q.where('team_main', null)
+			);
+
+			if (teamChannels.length) {
+				navigation.navigate('SelectListView', {
+					title: 'Leave_Team', room, data: teamChannels, infoText: 'Select_Team_Channels', nextAction: data => this.handleLeaveTeam(data), showAlert: () => this.showNoLeaveTeamAlert()
+				});
+			} else {
+				Alert.alert(
```
Comment (diegolmello) [inline app/views/RoomActionsView/index.js:475]: `showConfirmationAlert`

File: app/views/RoomActionsView/index.js [hunk 19]
```diff
@@ -568,9 +649,9 @@ class RoomActionsView extends React.Component {
 				<List.Section>
 					<List.Separator />
 					<List.Item
-						title='Leave_channel'
+						title={room.teamMain ? 'Leave' : 'Leave_channel'}
```
Comment (diegolmello) [inline app/views/RoomActionsView/index.js:652]: `Leave` is enough IMO

**Review by diegolmello — CHANGES_REQUESTED** [review]

**Review by diegolmello — APPROVED** [review]

**Pull request merged**